lib/datasets.go

```diff
@@ -1009,16 +1009,24 @@ func (m *DatasetMethods) Validate(p *ValidateParams, res *ValidateResponse) erro
 		return qrierr.New(ErrBadArgs, "please provide a dataset name, or a supply the --body and --schema or --structure flags")
 	}
 
-	ref, err := repo.ParseDatasetRef(p.Ref)
-	if err != nil && err != repo.ErrEmptyRef {
-		return err
-	}
-	err = repo.CanonicalizeDatasetRef(m.inst.repo, &ref)
-	if err != nil && err != repo.ErrEmptyRef {
-		if err == repo.ErrNotFound {
-			return fmt.Errorf("cannot find dataset: %s", ref)
+	fsiPath := ""
+	var err error
+	ref := dsref.Ref{}
+
+	// if there is both a bodyfilename and a schema/structure
+	// we don't need to resolve any references
+	if !(p.BodyFilename != "" && schemaFlagType != "") {
```

Comment:
The `Validate` code path, with its variety of options, is pretty confusing. Tests were very helpful here, and I'm not sure this is the right time, but a streamlined refactor that does a better job of illustrating the flow and minimizing the 'gotchya's' would be helpful here.

Reply:
So many negatives! For this condition, you can apply demorgan's law to make it read a little easier, and update the comment at the same time:

```
// If either bodyfile or schema is not provided, need to resolve the reference
if p.BodyFilename == "" || schemaFlagType == "" {
```